tests_e2e/playwright/navigator.py

```diff
@@ -0,0 +1,433 @@
+        self.page.get_by_role("button", name="Next").click()
+        self.page.get_by_role("textbox", name="Enter your password").fill(self.password)
+
+        self.page.wait_for_load_state("networkidle")
+        self.page.get_by_role("button", name="Next").click()
+
+        # let the page load
+        self.page.wait_for_load_state("networkidle")
+
+    def login_password(self) -> None:
+        """Go to a nebari deployment, login via Username/Password, and start
+        a new server.
+        """
+        logger.debug(">>> Sign in via Username/Password")
+        self.page.goto(self.nebari_url)
+        expect(self.page).to_have_url(re.compile(f"{self.nebari_url}*"))
+
+        self.page.get_by_role("button", name="Sign in with Keycloak").click()
+        self.page.get_by_label("Username").fill(self.username)
+        self.page.get_by_label("Password").click()
+        self.page.get_by_label("Password").fill(self.password)
+        self.page.get_by_role("button", name="Sign In").click()
+
+        # let the page load
+        self.page.wait_for_load_state("networkidle")
+
+    def start_server(self) -> None:
+        """Start a nebari server. There are several different web interfaces
+        possible in this process depending on if you already have a server
+        running or not. In order for this to work, wait for the page to load,
+        we look for html elements that exist when no server is running, if
+        they aren't visible, we check for an existing server start option.
+        """
+        # wait for the page to load
+        logout_button = self.page.get_by_text("Logout", exact=True)
+        logout_button.wait_for(state="attached")
+
+        # if server is not yet running
+        start_locator = self.page.get_by_role("button", name="Start My Server")
+        if start_locator.is_visible():
+            start_locator.click()
+
+            # select instance type (this will fail if this instance type is not
+            # available)
+            self.page.locator(f"#profile-item-{self.instance_name}").click()
+            self.page.get_by_role("button", name="Start").click()
+
+        else:
+            # if the server is already running
+            start_locator = self.page.get_by_role(
+                "button",
+                name="My Server",
+                exact=True,
+            )
+            start_locator.click()
+
+        # wait for server spinup
+        self.page.wait_for_url(
+            urllib.parse.urljoin(self.nebari_url, f"user/{self.username}/*"),
+            wait_until="networkidle",
+        )
+
+        # the jupyter page loads independent of network activity so here
+        # we wait for the File menu to be available on the page, a proxy for
+        # the jupyterlab page being loaded.
+        file_locator = self.page.get_by_text("File", exact=True)
+        file_locator.wait_for(
+            timeout=self.wait_for_server_spinup,
+            state="attached",
+        )
```

Comment:
I asked about this before. Could you explain why we need both conditions? My understanding is the following:

We wait for state to become `"attached"` **or** until the timeout. But what happens if it was actually the timeout that ended the waiting? IIUC, the code afterwards just assumes that the state is reached. But what if that never happens? Can't we get errors down the line that will be hard to debug because they are based on false assumptions?

Since we have the global timeout we are in no danger to run into an infinite wait here. Thus, unless I'm mistaken above, we should probably remove the timeout from `wait_for` if we have a way to determine if the locator reached the state we want.

Reply:
Ah, ok. That makes more sense. Meaning, this state has to be reached within the timeout frame and otherwise we error out, right? I'm ok with that then, but I don't think it is really necessary as we have the global timeout. Up to you if we want to keep it.